samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinatorFactory.java

```diff
@@ -40,22 +41,25 @@
   private static final String DEFAULT_JOB_NAME = "defaultJob";
 
   /**
-   * Method to instantiate an implementation of JobCoordinator
+   * Instantiates an {@link ZkJobCoordinator} using the {@link Config}.
    *
-   * @param config - configs relevant for the JobCoordinator TODO: Separate JC related configs into a "JobCoordinatorConfig"
-   * @return An instance of IJobCoordinator
+   * @param config zookeeper configurations required for instantiating {@link ZkJobCoordinator}
+   * @return An instance of {@link ZkJobCoordinator}
    */
   @Override
   public JobCoordinator getJobCoordinator(Config config) {
+    // TODO: Separate JC related configs into a "ZkJobCoordinatorConfig"
     MetricsRegistry metricsRegistry = new MetricsRegistryMap();
-    ZkUtils zkUtils = getZkUtils(config, metricsRegistry);
-    LOG.debug("Creating ZkJobCoordinator instance with config: {}.", config);
-    return new ZkJobCoordinator(config, metricsRegistry, zkUtils);
+    String jobCoordinatorZkBasePath = getJobCoordinationZkPath(config);
+    ZkUtils zkUtils = getZkUtils(config, metricsRegistry, jobCoordinatorZkBasePath);
+    MapConfig mapConfig = new MapConfig(config, ImmutableMap.of(ZkConfig.CONFIG_COORDINATOR_ZK_BASE_PATH, jobCoordinatorZkBasePath));
```

Comment:
I'm assuming this constructor is combining the two maps? If so, didn't we just get the jobCoordinatorZkBasePath from config? Do we need to add it back again?

Reply:
We construct the basePath(`getJobCoordinationZkPath`) by combining `appId`, `jobId`, `jobName` from the config. By combining the maps we add it to the configuration so that the layers beneath this doesn't have to construct the base path themselves(using the same logic).

Reply:
Not sure I understand what's happening here. Is the "derived" zkBasePath being written back to the config for the ZkJobCoordinator to use? Does that mean that `ZkConfig.CONFIG_COORDINATOR_ZK_BASE_PATH` is a derived config, and not user settable?

Reply:
What I meant to say was that if it's not exposed to users, should this be a separate parameter instead of an "internal" config key? I'd prefer to not wire internal properties through Config classes. We can discuss offline if I'm misunderstanding something.